### ISSUE TITLE
Downgrade piku requirement to python 3.4 

### ISSUE THREAD

**Issue by jsenin** [piku/piku#149]
Due debian jessie ships python3.4 and piku don't require any feature from python3.5, may we downgrade requirement to vesion 3.4 and let to run piku on old debian jessie boxes

https://packages.debian.org/es/jessie/python3
https://github.com/piku/piku/blob/master/piku.py#L7

**Comment by dwightmulcahy** [comment]
for reference of the differences... 
https://docs.python.org/3/whatsnew/3.5.html

**Comment by rcarmo** [comment]
I'm cool with this. I'll add tests for 3.4 as well.

**Comment by rcarmo** [comment]
Let's set a date for removing this, OK? We're now testing from 3.4 -> 3.8, and ideally I'd like to move to 3.6 as a baseline by... end of next year?

**Comment by rcarmo** [comment]
Hmmm. It turns out that we can't run 3.4 on GitHub Actions. Sorry, guys, but 3.5+ it is, otherwise I don't see how we can test this.

**Comment by chr15m** [comment]
@rcarmo I think backwards compatibility and not breaking things for the user are important in general. Updating the baseline means pushing work on to developers who use Piku. Of course there are limits, but if we *can* support an older version, I think we should.

I'm especially thinking of a situation where somebody has a Raspberry Pi in production in some embedded device that is not connected to the internet. One day they want to make some updates and they find that Piku no longer supports the Python version on their machine. So to make a change they have to upgrade the entire OS. This means pushing a new cost onto the user of Piku.

Rather, I think it is better to keep the costs to ourselves and maintain backwards compatibility where possible.

I understand the tradeoffs with security etc. but I like to think people are smart enough to negotiate those tradeoffs themselves, and if their particular case says that it's ok not to upgrade the OS then they shouldn't have to.

tl;dr: this video says it better than me: https://www.youtube.com/watch?v=V9ZC2CwkE0I

**Comment by chr15m** [comment]
Possible solution: lower the 3.5 requirement to 3.4 in the script, but say that the 3.4 version is not being unit tested by the infrastructure. People who wish to support 3.4 can push patches. What do you think?

**Comment by rcarmo** [comment]
Might work. The thing about 3.4 is that I don’t see it being supported anywhere these days, and Raspbian 9 already shipped with 3.5.3 (a fresh install I did yesterday comes with 3.7). So that means anyone who installed a Pi in the past year is already at 3.5, and that we’ll be at 3.8 within a year...

**Comment by chr15m** [comment]
Understood.

A bit of a data point:

![image](https://user-images.githubusercontent.com/67130/69616252-44cf3300-1071-11ea-8464-b8833a4d85b6.png)

 * 1.61 is Jessie = 33525 popcon installs
 * 1.64 is Stretch = 74751 popcon installs
 * 1.69 is Buster (stable) = 63321 popcon installs

As you can see Jessie still has an installed base about 50% the size of Stretch and Buster (but falling fast).

[Source](https://popcon.debian.org/).

**Comment by jsenin** [comment]
Interesting conversation... 

I think keeping old versions is a plus. Surely there are raspberries with versions based on Debian Jessie and they are not up to date or will be. The problem is how to test python3.4 support of automatic format since it is deprecated. So we can do as @chr15m says, pass tests over 3.5 and say that it supports 3.4 but it is untested or maybe pyenv can be used.

Anyway I think we do not use any feature that requires a version higher than 3.4, would be just for security reasons, or for using fstrings in 3.6 (@dwightmulcahy  :wink:)

These days I am playing to install piku in a Debian Jessie container and I am encountering these and other questions, that may can help to open the focus to older systems.

Ah! I just to remember that we should check python packages dependencies....

click-7.0  python>=2.7, python>=3.4
virtualenv-15.1.0 requires Python 2.6 or higher
uwsgi (master) python3>= 3.4

So, I think keep requirement to 3.4 should not be a problem

**Comment by rcarmo** [comment]
I've been trying to find equivalent statistics for Ubuntu and Raspbian, but there's nothing directly comparable (although Raspbian should have a similar curve to Debian mainline, except for the few weeks of their early Buster adoption...).

Anyway, getting down to brass tacks, since technically we don't, in fact, have any issues with 3.4.

Can we just issue a warning if running below 3.5 and exit if running under 3.4, and move those goalposts on a semi-yearly basis? 

That would mean deprecating 3.4 completely by Summer 2020, and targeting 3.6 at the same time (issuing a warning for 3.5), etc.

Bear in mind that this also ties in with plugins (once we get there).

**Comment by chr15m** [comment]
> deprecating 3.4 completely by Summer 2020, and targeting 3.6 at the same time (issuing a warning for 3.5), etc.

I'm a bit wary of this. Would prefer to have this informed by actual usage rather than an time based rule.

What do you think about a percentage cutoff? So if the data looks like less than x% of users are on a Python version then we can deprecate.

**Comment by chr15m** [comment]
> move those goalposts on a semi-yearly basis

What are the reasons in favour of doing this? What benefit does it bring to the users of Piku and to us, the maintainers?

**Comment by rcarmo** [comment]
Well, I’d much prefer targeting new installs and current-1 of Ubuntu, really, and AFAIK there aren’t exactly thousands of people using Piku at the moment, so testing and supporting current Raspbian would be my first priority, with Ubuntu second (because that is what I see in broader use than Debian in my professional context...)

**Comment by jsenin** [comment]
I just have an idea that may like you...
It's very easy

```
try:
    from sys import version_info
    assert version_info >= (3, 4)
    if version_info < (3, 5):
      print("Warning, running piku python 3.4 might result in some unexpected behaviour, please upgrade to python 3.5 or greather")
except AssertionError:
    exit("Piku requires Python 3.5 or above")

```
I thinks this is a diplomatic solution :D

**Comment by rcarmo** [comment]
Yeah, almost exactly what I had in mind.

**Comment by rcarmo** [comment]
I'm going to close this as we'll soon have to move to 3.6-3.7 as a minimum (since there are already distros with 3.8) and it would be nice to use f-strings.
